core/iwasm/common/wasm_shared_memory.c

```diff
@@ -489,17 +497,27 @@ wasm_runtime_atomic_notify(WASMModuleInstanceCommon *module, void *address,
     WASMModuleInstance *module_inst = (WASMModuleInstance *)module;
     uint32 notify_result;
     AtomicWaitInfo *wait_info;
+    WASMSharedMemNode *node;
+    bool out_of_bounds;
 
     bh_assert(module->module_type == Wasm_Module_Bytecode
               || module->module_type == Wasm_Module_AoT);
 
-    if ((uint8 *)address < module_inst->memories[0]->memory_data
-        || (uint8 *)address + 4 > module_inst->memories[0]->memory_data_end) {
+    node = search_module((WASMModuleCommon *)module_inst->module);
+    os_mutex_lock(&node->shared_mem_lock);
```

Comment:
this line is creating problems in the spec tests, I'm checking why

Reply:
> This line is creating problems in the spec tests, I'm checking why.
> 
> [EDIT]: The problem is that, according to the specification, `atomic.notify` is supposed to also work without shared memory (in contrast with `atomic.wait`). But, without shared memory, `node->shared_mem_lock` is not initialized, making the program crash.

The `node->shared_mem_lock` is initialized when node is created, I think what you mean should be that `node` may be NULL when `atomic.notify`. How about adding checks for it:
```C
    if (node)
        os_mutex_lock(&node->shared_mem_lock);
    ut_of_bounds =
        ((uint8 *)address < module_inst->memories[0]->memory_data
         || (uint8 *)address + 4 > module_inst->memories[0]->memory_data_end);
    if (node)
        os_mutex_unlock(&node->shared_mem_lock);
```
```

Reply:
Right! I was mislead since I was using an assert to check that `node` was not null without realizing that spec tests were executed in release mode. Thanks